### DISCUSSION TITLE
how to make a license key schema using prisma that can generate promotional code like 25% off, 50% off, or 100% off?

### DISCUSSION THREAD

**Discussion by deadcoder0904** [prisma/prisma#13431] [Q&A]
my prisma schema looks like:

```prisma
model User {
  id       String  @id @default(cuid())
  email    String? @unique
  stripeId String  @unique

  product Product?

  @@map("users")
}

model Product {
  id                   String @id @default(cuid())
  licenseKey           String
  /// no. of license keys generated
  generatedLicenseKeys Int    @default(1)
  /// no. of license keys used
  usedLicenseKeys      Int    @default(0)
  /// stored in cents
  totalSum             Int    @default(9700)

  user   User   @relation(fields: [userId], references: [id])
  userId String @unique

  @@map("products")
}
```

when a user pays full price for the product, it generates a license for the user & points back to the email.

but i also want a way to create license keys from admin panel where it doesn't point back to any email.

like i want 20 licenses with promotional code `CHRISTMAS_SALE_20_OFF` & i want them to point to no email while creating. it will be connected to emails when someone uses the code.

i know i need a 1-to-many relation but do i need another table? can anyone help me out here?

**Comment by pantharshit00** [comment]
Hello @deadcoder0904 , our apologies for the delay in getting back to you! Due to the large number of questions in our community channels, unfortunately this slipped through the cracks. It's very important for us to address your question, so let us know in a comment if this is still an open discussion. 

I would model this as following:
```prisma
model User {
  id       String  @id @default(cuid())
  email    String? @unique
  stripeId String  @unique

  product  Product?
  licenses License[]

  @@map("users")
}

model License {
  id          String    @id @default(cuid())
  licenseKey  String    @unique
  user        User?     @relation(fields: [userId], references: [id])
  userId      String
  product     Product   @relation(fields: [productId], references: [id])
  productId   String
  price       BigInt
  promoCode   Promocode @relation(fields: [promocodeId], references: [id])
  discount    BigInt
  promocodeId String
}

model Promocode {
  id        String    @id @default(cuid())
  code      String    @unique
  value     BigInt
  isPercent Boolean
  licenses   License[]
}

model Product {
  id                   String @id @default(cuid())
  licenseKey           String
  /// no. of license keys generated
  generatedLicenseKeys Int    @default(1)
  /// no. of license keys used
  usedLicenseKeys      Int    @default(0)
  /// stored in cents
  totalSum             Int    @default(9700)

  user     User      @relation(fields: [userId], references: [id])
  userId   String    @unique
  licenses License[]

  @@map("products")
}
```

Here `License` model has user as optional and you will set the user when they will activate the license. The `Promocode` model here stores the `code` and the `value`. You will use these values to compute price in the application.

**Reply by vladi-stevanovic** [reply]
:wave: Hello @deadcoder0904 , I’m going through pending GitHub Discussions to check if you need any help. Is this issue still open?

Please let us know if we can investigate it further or if you've already solved it, we'd love to know how!

It would be super useful for other users facing this problem! :raised_hands:

**Reply by deadcoder0904** [reply]
@pantharshit00 @vladi-stevanovic i have made it work on another project... i'll try @pantharshit00's solution & compare since i'm launching something similar this month so i'll let you know by then :)

**Reply by vladi-stevanovic** [reply]
Sounds good, thank you for the follow up @deadcoder0904, and keep up posted!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
